### PR TITLE
Add spew debugging

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard jhoward@microsoft.com

@cpuguy83 PTAL as I know you like your debugging stuff :smile: 

Adding some extremely useful additional debugging when reacting to `SIGUSR1` (Linux) or the debugging event being signalled on Windows. I've debugged so many deadlocks now, and while the stacks are useful, getting the daemon data-structures out as well is likely to prove invaluable. 

This PR uses https://github.com/davecgh/go-spew to do the reflection/pretty-printing etc.

Note that NO LOCKS ARE TAKEN. Hence the data-structure could potentially be changing during the dump and cause a panic. It is very deliberate to not take any locks, as if the daemon is deadlocked, that would cause the dumper routine to potentially deadlock too. For that reason, there's an explicit `recover` added in a goroutine to ensure that the daemon does not panic.

Here's a sample output on Windows:

```
PS C:\control> $p=(get-process dockerd).id;docker-signal -pid $p
Daemon signalled successfully. Examine its output for stacks
```

```
time="2016-10-12T16:00:40.720524700-07:00" level=info msg="API listen on //./pipe/docker_engine"
time="2016-10-12T16:00:47.961900400-07:00" level=info msg="goroutine stacks written to C:\\control\\goroutine-stacks.log"
time="2016-10-12T16:00:48.212779200-07:00" level=info msg="daemon dump written to C:\\control\\daemon-data.log"
```

```
PS C:\control> type .\daemon-data.log
=== BEGIN daemon dump ===
2016-10-12 16:00:47.9638575 -0700 PDT
(*daemon.Daemon)(0xc0423bc540)({
 ID: (string) (len=59) "W4GO:UXRR:PWJE:CLWX:3K6O:PFGX:I2KN:WVRR:V4G4:NKAR:QOBY:AEW7",
 repository: (string) (len=21) "C:\\control\\containers",
 containers: (*container.memoryStore)(0xc042155ee0)({
  s: (map[string]*container.Container) (len=1) {
   (string) (len=64) "3bcc1fd2c454fa545ce23ff7b9976a27c916fa58ac450079e6e9aba79b991055": (*container.Container)(0xc042423d40)()
  },
  RWMutex: (sync.RWMutex) {
   w: (sync.Mutex) {
    state: (int32) 0,
    sema: (uint32) 0
   },
   writerSem: (uint32) 0,
   readerSem: (uint32) 0,
   readerCount: (int32) 0,
   readerWait: (int32) 0
  }
 }),
 execCommands: (*exec.Store)(0xc042155f00)({
  commands: (map[string]*exec.Config) {
  },
  RWMutex: (sync.RWMutex) {
   w: (sync.Mutex) {
    state: (int32) 0,
    sema: (uint32) 0
   },
   writerSem: (uint32) 0,
   readerSem: (uint32) 0,
   readerCount: (int32) 0,
   readerWait: (int32) 0
  }
 }),
 referenceStore: (*reference.store)(0xc04203cb40)({
  mu: (sync.RWMutex) {
   w: (sync.Mutex) {
    state: (int32) 0,
    sema: (uint32) 0
   },
   writerSem: (uint32) 0,
   readerSem: (uint32) 0,
   readerCount: (int32) 0,
   readerWait: (int32) 0
  },
  jsonPath: (string) (len=48) "C:\\control\\image\\windowsfilter\\repositories.json",
  Repositories: (map[string]reference.repository) (len=4) {
   (string) (len=7) "busybox": (reference.repository) (len=1) {
    (string) (len=14) "busybox:latest": (digest.Digest) (len=71) sha256:eaaaeb371ce6803f3a54287767df5bbbd3b2150b385cacdf202b81419640c62f
   },
   (string) (len=10) "nanoserver": (reference.repository) (len=2) {
    (string) (len=25) "nanoserver:10.0.14393.206": (digest.Digest) (len=71) sha256:57be450d96ed8aefd3ad7b5ce32ec5d370debad0ae88a2182a57bff052055cf7,
    (string) (len=17) "nanoserver:latest": (digest.Digest) (len=71) sha256:57be450d96ed8aefd3ad7b5ce32ec5d370debad0ae88a2182a57bff052055cf7
   },
   (string) (len=4) "psql": (reference.repository) (len=1) {
    (string) (len=11) "psql:latest": (digest.Digest) (len=71) sha256:f594941fa618d5fb8e46495ff5b31a804b6b58f0d88fd9bc5531402a51c53c96
   },
   (string) (len=17) "windowsservercore": (reference.repository) (len=2) {
    (string) (len=32) "windowsservercore:10.0.14393.206": (digest.Digest) (len=71) sha256:d754b1c586b2c9a96baea0d9367b43c315a8b5c181284fa8baa8c8286a4183d8,
    (string) (len=24) "windowsservercore:latest": (digest.Digest) (len=71) sha256:d754b1c586b2c9a96baea0d9367b43c315a8b5c181284fa8baa8c8286a4183d8
   }
  },
  referencesByIDCache: (map[digest.Digest]map[string]reference.Named) (len=4) {
   (digest.Digest) (len=71) sha256:eaaaeb371ce6803f3a54287767df5bbbd3b2150b385cacdf202b81419640c62f: (map[string]reference.Named) (len=1) {
    (string) (len=14) "busybox:latest": (*reference.taggedRef)(0xc0423ce380)(busybox:latest)
   },
   (digest.Digest) (len=71) sha256:57be450d96ed8aefd3ad7b5ce32ec5d370debad0ae88a2182a57bff052055cf7: (map[string]reference.Named) (len=2) {
    (string) (len=25) "nanoserver:10.0.14393.206": (*reference.taggedRef)(0xc0423ce1c0)(nanoserver:10.0.14393.206),
    (string) (len=17) "nanoserver:latest": (*reference.taggedRef)(0xc0423ce210)(nanoserver:latest)
   },
   (digest.Digest) (len=71) sha256:f594941fa618d5fb8e46495ff5b31a804b6b58f0d88fd9bc5531402a51c53c96: (map[string]reference.Named) (len=1) {
    (string) (len=11) "psql:latest": (*reference.taggedRef)(0xc0423ce280)(psql:latest)
   },
   (digest.Digest) (len=71) sha256:d754b1c586b2c9a96baea0d9367b43c315a8b5c181284fa8baa8c8286a4183d8: (map[string]reference.Named) (len=2) {
    (string) (len=32) "windowsservercore:10.0.14393.206": (*reference.taggedRef)(0xc0423ce2d0)(windowsservercore:10.0.14393.206),
    (string) (len=24) "windowsservercore:latest": (*reference.taggedRef)(0xc0423ce320)(windowsservercore:latest)
   }
  }
 }),
 downloadManager: (*xfer.LayerDownloadManager)(0xc0423c3440)({
  layerStore: (*layer.layerStore)(0xc042007e40)({
   store: (*layer.fileMetadataStore)(0xc0423ce6a0)({
    root: (string) (len=38) "C:\\control\\image\\windowsfilter\\layerdb"
   }),
   driver: (*windows.Driver)(0xc0423c6c30)(windowsfilter),
   layerMap: (map[layer.ChainID]*layer.roLayer) (len=11) {
    (layer.ChainID) (len=71) sha256:ff3afc74726160ea24b664529ad3ab983b7a572c82742e00c53ad409bcfa643d: (*layer.roLayer)(0xc0423e0460)({
     chainID: (layer.ChainID) (len=71) sha256:ff3afc74726160ea24b664529ad3ab983b7a572c82742e00c53ad409bcfa643d,
     diffID: (layer.DiffID) (len=71) sha256:a6e513bc458471199d83f8f6c2f0593f2f505ef451133cfc4734628abe4fef41,
     parent: (*layer.roLayer)(0xc0423e0140)({
      chainID: (layer.ChainID) (len=71) sha256:0ce98b97c17e1007db26ace432c35229aaa7d48ab6632989e6dd44e588c871a9,
      diffID: (layer.DiffID) (len=71) sha256:c3f6e5273094be71a347dc0dda60b72e60415b9c61b60f6dadf8aa89bbc0c9d7,
      parent: (*layer.roLayer)(0xc0423e01e0)({
       chainID: (layer.ChainID) (len=71) sha256:e654f3dbb45c05c83942df4d0f67d60f29d3182c923cee7a170c7b7f12baa869,
       diffID: (layer.DiffID) (len=71) sha256:0fe6c66aab9eed3388ed31a88faa4d1e2f4f4f194cc8d396d45e1f71005f32e2,
       parent: (*layer.roLayer)(0xc0423e0320)({
        chainID: (layer.ChainID) (len=71) sha256:2c08913762fe083474c0a9eec5405fdd99203472ccfcfbedecaf768612b68332,
        diffID: (layer.DiffID) (len=71) sha256:2c08913762fe083474c0a9eec5405fdd99203472ccfcfbedecaf768612b68332,
        parent: (*layer.roLayer)(<nil>),
        cacheID: (string) (len=64) "c38ac738bb8c6e079718266f60a4223aab5848214b5cd6caeadb30c6f11e2a89",
        size: (int64) 662468910,
        layerStore: (*layer.layerStore)(0xc042007e40)(<already shown>),
        descriptor: (distribution.Descriptor) {
         MediaType: (string) (len=57) "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
         Size: (int64) 242834639,
         Digest: (digest.Digest) (len=7) -diffid,
         URLs: ([]string) (len=1 cap=4) {
          (string) (len=14) "layer_is_local"
         }
        },
        referenceCount: (int) 5,
        references: (map[layer.Layer]struct {}) (len=1) {
         (*layer.referencedCacheLayer)(0xc042028378)({
          roLayer: (*layer.roLayer)(0xc0423e0320)(<already shown>)
         }): (struct {}) {
         }
        }
       }),
       cacheID: (string) (len=64) "d74115d4724657f657d0408f01c5d899fe0d414dd847b605e37918fbb2933720",
       size: (int64) 533504,
       layerStore: (*layer.layerStore)(0xc042007e40)(<already shown>),
       descriptor: (distribution.Descriptor) {
        MediaType: (string) "",
        Size: (int64) 0,
        Digest: (digest.Digest) ,
        URLs: ([]string) <nil>
       },
...<snipped>
=== END daemon dump dump ===
PS C:\control>
```

### CODE REVIEW

**Comment by jrick** [comment]
I don't follow docker dev but I saw this linked from the go-spew issue, so here I am.

> Note that NO LOCKS ARE TAKEN. Hence the data-structure could potentially be changing during the dump and cause a panic.

This worries me a lot.  Data races are a violation of the Go memory model, and create the same situation of undefined behavior you might expect to see in a C++ program.  This can result in the entire program's execution, from start to finish, to be undefined (even before UB "happens").

The most reliable way to inspect the memory of a deadlocked program (at least in Unix) would be to dump a core.  I'm sure similar methods exist in Windows.

**Comment by jhowardmsft** [comment]
I'm sure this is safe having artificially forced a few deadlocks in testing it. Unfortunately no good tools really exist to do the equiv of a core dump that can examine go structures on Windows. Worst case, the daemon crashes. But if it's already deadlocked, that's the least concern. On Windows it may well be deadlocked due to a win32 call not returning (several past examples) while the daemon is holding a lock which blocks other threads. In those cases, this is also perfectly safe as nothing will be changing under it. Just more data to aide debugging.

**Comment by jhowardmsft** [comment]
And to add, the added code also has a recover go function if there is a panic, and I've verified that that copes by forcing a nil pointer indirection. The spew code should just be a reader, not a writer.

**Comment by thaJeztah** [comment]
ping @LK4D4 @stevvooe ptal

**Comment by stevvooe** [comment]
This looks very useful. I wish there was a standard Go way to do this. Is there something we're doing that makes it hard to use [delve](https://github.com/derekparker/delve)?

I am slightly worried about the lack of locking here. Is the daemon memory safe after doing a debugging run or should we exit? Considering this is for debugging deadlocks, that may not be a real concern.

File: daemon/debugtrap.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+	// Ensure we recover from a panic as we are doing this without any locking
+	defer func() {
+		if r := recover(); r != nil {
+		}
+	}()
+
+	if root == "" {
+		logrus.Infof("=== BEGIN daemon dump ===\n%s\n=== END daemon dump ===", spew.Sdump(d))
+	} else {
+		path := filepath.Join(root, "daemon-data.log")
```
Comment (jhowardmsft) [inline daemon/debugtrap.go]: Yes, it does

**Comment by jhowardmsft** [comment]
Delve is an option if you can keep the machine for further investigation. That's not always the case for me over the past few months. However, you need source access on the machine, and delve or GDB are difficult to use on a nano server host, almost next to impossible. 

I think this is safe and the benefits outweigh the risks. I would also rather keep the daemon running as that does also leave the option of subsequent attaching using delve or GDB. Given that this is a reader, it shouldn't have any further adverse effects.

**Comment by stevvooe** [comment]
@jhowardmsft Fair enough.

@diogomonica @NathanMcCauley Let's do a security review on this one.

File: daemon/debugtrap.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+		f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
```
Comment (diogomonica) [inline daemon/debugtrap.go]: why `0666`?

File: daemon/debugtrap.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/Sirupsen/logrus"
+	"github.com/davecgh/go-spew/spew"
+)
+
+// dumpDaemon is a helper function to dump the daemon data-structures
+// for debugging purposes. NO LOCKS ARE TAKEN! (Useful for deadlocks).
+func (d *Daemon) dumpDaemon(root string) {
+	// Ensure we recover from a panic as we are doing this without any locking
+	defer func() {
+		if r := recover(); r != nil {
```
Comment (AkihiroSuda) [inline daemon/debugtrap.go]: why this `if` lacks body statements?

File: daemon/debugtrap.go (hunk 2 above)
Comment (jhowardmsft) [inline daemon/debugtrap.go]: Carry over from writing the stack trace dump which does the same. It's effectively a no-op on Windows anyway as there's no mode flags in the same way as for Linux.

File: daemon/debugtrap.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+		logrus.Infof("=== BEGIN daemon dump ===\n%s\n=== END daemon dump ===", spew.Sdump(d))
+	} else {
+		path := filepath.Join(root, "daemon-data.log")
+		f, err := os.OpenFile(path, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
+		if err != nil {
+			logrus.Warnf("Could not open %s to write the daemon dump: %v", path, err)
+			return
+		}
+		defer f.Close()
+		f.WriteString("=== BEGIN daemon dump ===\n")
```
Comment (jhowardmsft) [inline daemon/debugtrap.go]: Not sure I follow...?

Comment (jhowardmsft) [inline daemon/debugtrap.go]: Ah, was reading on my phone. More obvious on a computer 😇. Sure will update as needs rebasing anyway.

**Comment by jhowardmsft** [comment]
Rebased. Can we move this forward to code review? I'd really like to get this in for 1.13. @thaJeztah

File: daemon/debugtrap.go (hunk 3 above)
Comment (cpuguy83) [inline daemon/debugtrap.go]: But you'd just need the `recover()`, no if block.

**Comment by jhowardmsft** [comment]
@cpuguy83 Yes, you're right. Removed the if block. I was blindly copying an example.

**Comment by jhowardmsft** [comment]
Rebased

**Comment by thaJeztah** [comment]
ping @cpuguy83 @stevvooe PTAL

**Comment by jhowardmsft** [comment]
Rebased again.

**Comment by thaJeztah** [comment]
ping @AkihiroSuda perhaps you can have another look, to see if we can get this moving?

**Comment by stevvooe** [comment]
LGTM

I am not sure of a better way forward here. We should reserve the right to remove this at any time when the debuggers can do core analysis.

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by AkihiroSuda** [comment]
@thaJeztah LGTM (sorry for being late 😅)

**Comment by thaJeztah** [comment]
@AkihiroSuda no worries!

**Comment by tonistiigi** [comment]
In linux this debug file is `6.6 MB`, over 100000 lines and contains a precalculated array of 65536 IP addresses for an empty daemon. Surely we don't need all that data and it just makes it harder to move around that file. Of course, we should also see why we are allocation so many things.

**Comment by thaJeztah** [comment]
@tonistiigi can you open an issue for tracking that?

**Pull request merged**